### PR TITLE
Unique message id mapper contract tests

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Second commit name should be " Add MessageIdMapper contract tests"

(You got a extra "m")

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMapperProvider.java [hunk 1]
```diff
@@ -129,4 +138,9 @@ public AnnotationMapper createAnnotationMapper() throws MailboxException {
     public List<Capabilities> getNotImplemented() {
         return ImmutableList.of();
     }
+
+    @Override
+    public MessageUid generateMessageUid() {
+        return MessageUid.of(RandomUtils.nextLong());
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMapperProvider.java]: That's not correct, Uids should be monotically incrising per mailbox.

Maybe we should write a wrapper around a Int we increase on each call to next, that would be more correct...

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/mail/InMemoryMapperProvider.java [hunk 2]
```diff
@@ -48,6 +58,13 @@ public InMemoryId generateId() {
     }
 
     @Override
+    public MessageUid generateMessageUid() {
+        long currentUid = nextUid;
```
Comment (chibenwa) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/mail/InMemoryMapperProvider.java]: Do the same in Cassandra ;-)

And maybe extract it to an other class

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+import org.junit.Rule;
+import org.junit.rules.ExpectedException;
+import org.xenei.junit.contract.Contract;
+import org.xenei.junit.contract.ContractTest;
+import org.xenei.junit.contract.IProducer;
+
+import com.google.common.collect.ImmutableList;
+
+@Ignore
+@Contract(MapperProvider.class)
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: Mphhh, time to migrate away from contact testing, no ?

 - It won't run on CI
 - Can't run tests individually

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+        List<MailboxId> mailboxes = sut.findMailboxes(message1.getMessageId());
+        assertThat(mailboxes).containsOnly(benwaInboxMailbox.getMailboxId());
+    }
+
+    @ContractTest
+    public void findMailboxesShouldReturnTwoMailboxesWhenMessageExistsInTwoMailboxes() throws MailboxException {
+        saveMessages();
+
+        SimpleMailboxMessage message1InOtherMailbox = SimpleMailboxMessage.copy(benwaWorkMailbox.getMailboxId(), message1);
+        message1InOtherMailbox.setUid(mapperProvider.generateMessageUid());
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: And you don't set ModSeq ?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+
+        MessageId messageId = message1.getMessageId();
+        sut.setFlags(new Flags(Flag.ANSWERED), MessageManager.FlagsUpdateMode.REMOVE, messageId);
+
+        List<MailboxMessage> messages = sut.find(ImmutableList.of(messageId), MessageMapper.FetchType.Body);
+        assertThat(messages).hasSize(2);
+        assertThat(messages.get(0).isAnswered()).isTrue();
+        assertThat(messages.get(1).isAnswered()).isTrue();
+    }
+
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: Verify also modseq increment upon setFlags (It's mapper reponsibity to increment it on each retry...)

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+    private void saveMessages() throws MailboxException {
+        addMessageAndSetModSeq(benwaInboxMailbox, message1);
+        addMessageAndSetModSeq(benwaInboxMailbox, message2);
+        addMessageAndSetModSeq(benwaInboxMailbox, message3);
+        addMessageAndSetModSeq(benwaWorkMailbox, message4);
+    }
+
+    private void addMessageAndSetModSeq(Mailbox mailbox, MailboxMessage message) throws MailboxException {
+        messageMapper.add(mailbox, message);
+        message1.setModSeq(messageMapper.getHighestModSeq(mailbox));
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: With this all your messages will get the same modseq ^^

You might want to ask for ModSeqProvider through MapperProvider (easy to get it), and call it's next modseq.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+    public ExpectedException expectedException = ExpectedException.none();
+
+    private final static char DELIMITER = ':';
+    private static final int BODY_START = 16;
+    private final static long UID_VALIDITY = 42;
+
+    private IProducer<T> producer;
+    private MessageMapper messageMapper;
+    private MailboxMapper mailboxMapper;
+    private MessageIdMapper sut;
```
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: What does it mean "sut"?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+        message1InOtherMailbox.setUid(mapperProvider.generateMessageUid());
+        sut.save(message1InOtherMailbox);
+
+        MessageId messageId = message1.getMessageId();
+        sut.delete(messageId);
+
+        List<MailboxId> mailboxes = sut.findMailboxes(messageId);
+        assertThat(mailboxes).isEmpty();
+    }
+
```
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: Do we have any test case which delete message should
- Remove message
- Remove all MailboxMessage?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java (hunk 7 above)
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: Suite under test ;-)

**Comment by aduprat** [comment]
I will rename the commit & add ticket number when the PR will be approved, just before merge.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageIdMapper.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+    List<MailboxId> findMailboxes(MessageId messageId);
+
+    void save(MailboxMessage mailboxMessage) throws MailboxNotFoundException, MailboxException;
+
+    void delete(MessageId messageId);
+
+    void delete(MessageId messageId, List<MailboxId> mailboxIds);
+
+    Map<MailboxId, UpdatedFlags> setFlags(Flags newState, MessageManager.FlagsUpdateMode updateMode, MessageId messageId) throws MailboxException;
```
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageIdMapper.java]: messageId as first parameter?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 10]
```diff
@@ -0,0 +1,413 @@
...
+import com.google.common.collect.ImmutableList;
+
+@Ignore
+@Contract(MapperProvider.class)
+public class MessageIdMapperTest<T extends MapperProvider> {
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+
+    private final static char DELIMITER = ':';
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: dot is a more usual delimiter

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java (hunk 3 above)
Comment (aduprat) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: Ok, i've added the UNIQUE_MESSAGE_ID MapperProvider capability.
I wish to keep contract tests in order to stay consistent with other mappers tests.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 11]
```diff
@@ -0,0 +1,413 @@
...
+        sut.save(message1InOtherMailbox);
+
+        List<MailboxId> mailboxes = sut.findMailboxes(message1.getMessageId());
+        assertThat(mailboxes).containsOnly(benwaInboxMailbox.getMailboxId(), benwaWorkMailbox.getMailboxId());
+    }
+
+    @ContractTest
+    public void saveShouldSaveAMessage() throws Exception {
+        message1.setUid(mapperProvider.generateMessageUid());
+        sut.save(message1);
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: I'm not sure: it is save that should allocate an uid or do we need to set it before as you do?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 12]
```diff
@@ -0,0 +1,413 @@
...
+        notPersistedMailbox.setMailboxId(mapperProvider.generateId());
+        SimpleMailboxMessage message = createMessage(notPersistedMailbox, "Subject: Test \n\nBody\n.\n", BODY_START, new PropertyBuilder());
+        message.setUid(mapperProvider.generateMessageUid());
+
+        expectedException.expect(MailboxNotFoundException.class);
+        sut.save(message);
+    }
+
+    @ContractTest
+    public void saveShouldSaveMessageAnotherIndicesWhenMessageAlreadyInMailbox() throws Exception {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: I don't understand the name of this test

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 13]
```diff
@@ -0,0 +1,413 @@
...
+
+        MessageId messageId = message1.getMessageId();
+        sut.delete(messageId, ImmutableList.of(benwaInboxMailbox.getMailboxId(), benwaWorkMailbox.getMailboxId()));
+
+        List<MailboxId> mailboxes = sut.findMailboxes(messageId);
+        assertThat(mailboxes).isEmpty();
+    }
+
+    @ContractTest
+    public void setFlagsShouldReturnUpdatedFlagsWhenMessageIsInOneMailbox() throws Exception {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: A test is missing when message id does not exist.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java (hunk 8 above)
Comment (aduprat) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: . Remove message: we are not removing them yet, only indices.
. remove all MailboxMessage: i don't understand which kind of test you are waiting for.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java (hunk 10 above)
Comment (aduprat) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: Also changed on MessageMapperTest & MailboxMapperTest

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java (hunk 11 above)
Comment (aduprat) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: Setting attributes on MailboxMessage should be done before, as MailboxMessage has to be seen as immutable.

**Comment by chibenwa** [comment]
test this please

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 14]
```diff
@@ -331,6 +331,19 @@ public void setFlagsShouldReturnUpdatedFlagsWhenMessageIsInOneMailbox() throws E
...
+    public void setFlagsShouldReturnEmptyWhenMailboxIdsIsEmpty() throws Exception {
+        message1.setUid(mapperProvider.generateMessageUid());
+        message1.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message1);
+
+        MessageId messageId = message1.getMessageId();
+        Flags newFlags = new Flags(Flag.ANSWERED);
+        Map<MailboxId, UpdatedFlags> flags = sut.setFlags(messageId, ImmutableList.<MailboxId> of(), newFlags, MessageManager.FlagsUpdateMode.REMOVE);
+
+        assertThat(flags).isEmpty();
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: Can you check that the original message has not been modified too please?
And can you add some tests with several mailboxes?

Comment (aduprat) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: 1/ Ok
2/ Such tests already exists

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 15]
```diff
@@ -0,0 +1,509 @@
...
+        message1.setModSeq(modSeq);
+        sut.save(message1);
+
+        MessageId messageId = message1.getMessageId();
+        Flags newFlags = new Flags(Flag.ANSWERED);
+        sut.setFlags(messageId, ImmutableList.<MailboxId> of(), newFlags, MessageManager.FlagsUpdateMode.REMOVE);
+
+        List<MailboxMessage> messages = sut.find(ImmutableList.of(messageId), MessageMapper.FetchType.Body);
+        assertThat(messages).hasSize(1);
+        assertThat(messages.get(0).getModSeq()).isGreaterThan(modSeq);
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: Name of the test is setFlagsShouldNotModifyModSeqWhenMailboxIdsIsEmpty

Shouldn't it be isEqualTo ?

**Comment by aduprat** [comment]
test this please

**Pull request merged**